### PR TITLE
Recoded collisions.

### PR DESCRIPTION
**Pull request by Zly-u**
New collision system that works for both curving and static walls.
Haven't been done for Custom walls though, comming soon`tm` but for now the standard method is okay.

### CODE REVIEW

File: include/SSVOpenHexagon/Components/CPlayer.hpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #pragma once
 
 #include "SSVOpenHexagon/Utils/Ticker.hpp"
+#include "CWall.hpp"
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Components/CPlayer.hpp:8]: We don't need a physical dependency on `CWall.hpp` here, as we are only using `CWall` by reference (which does not require its definition). A [forward declaration](https://stackoverflow.com/questions/4757565/what-are-forward-declarations-in-c) would suffice. See line 18 in this file for an example. Reducing physical dependencies avoids dependency cycles and speeds up compilation time.

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 2]
```diff
@@ -19,35 +19,29 @@
 #include <SFML/System/Vector2.hpp>
 #include <SFML/Graphics/Color.hpp>
 
-namespace hg
-{
+namespace hg {
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:22]: For the future, avoid changing the formatting of code. It makes the diffs bigger than they need to be. I would recommend using `clang-format` using the `.clang-format` style available in the repository root.

File: src/SSVOpenHexagon/Online/Online.cpp [hunk 3]
```diff
@@ -516,7 +516,7 @@ void initializeValidators(HGAssets& mAssets)
 
         HG_LO_VERBOSE("hg::Online::initializeValidators")
             << "Finished initializing validators...\n";
-    */
+    //*/
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Online/Online.cpp:519]:
```suggestion
    */
```

Comment (SuperV1234) [reply src/SSVOpenHexagon/Online/Online.cpp:519]: Mistake?

File: src/SSVOpenHexagon/Core/HGUpdate.cpp [hunk 4]
```diff
@@ -266,6 +262,23 @@ void HexagonGame::update(ssvu::FT mFT)
         fpsWatcher.update();
     }
 }
+
+void HexagonGame::updateWalls(ssvu::FT mFT){
+    const auto playerPos{player.getPosition()};
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Core/HGUpdate.cpp:267]: Unused?

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 5]
```diff
@@ -136,17 +127,70 @@ void CPlayer::drawDeathEffect(HexagonGame& mHexagonGame)
...
+    if (!Config::getInvincible()) { dead = true; }
+    if (mEffect) { mHexagonGame.death(); }
+
+    ssvs::moveTowards(lastPos, ssvs::zeroVec2f, 5 * mHexagonGame.getSpeedMultDM());
+
+    pos = lastPos;
+}
+
+void CPlayer::push(HexagonGame& mHexagonGame, CWall& wall){
+    const auto curveData = wall.getCurve();
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:159]:
```suggestion
    const auto& curveData = wall.getCurve();
```

No need to copy here.

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 6]
```diff
@@ -136,17 +127,70 @@ void CPlayer::drawDeathEffect(HexagonGame& mHexagonGame)
...
+    const int movement{mHexagonGame.getInputMovement()};
+
+    const float radius{mHexagonGame.getRadius()};
+
+    unsigned int maxAttempts;
+    if (curveDir != 0){
+        maxAttempts = std::abs(curveData.speed+speed*(movement*curveDir));
+    } else {
+        maxAttempts = speed;
+    }
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:170]:
```suggestion
    const unsigned int maxAttempts = curveDir != 0 
        ? std::abs(curveData.speed+speed*(movement*curveDir)) : speed;
```

Comment (SuperV1234) [reply src/SSVOpenHexagon/Components/CPlayer.cpp:170]: This is good because it allows us to mark `maxAttempts` as `const`, so we know it won't change in the rest of the function

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 7]
```diff
@@ -136,17 +127,70 @@ void CPlayer::drawDeathEffect(HexagonGame& mHexagonGame)
...
+    unsigned int attempt = 0;
+    float pushDir;
+    if (curveDir != 0){
+        const int reverseDir = ((movement != 0) ? -movement : 1)*curveDir;
+        pushDir = ssvu::toRad(1.f)*curveDir*reverseDir;
+    } else {
+        pushDir = ssvu::toRad(1.f) * -movement;
+    }
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:178]: `const` would be nice here as well, but the logic is too complicated to put it in a ternary operator. It would be unreadable. We can use the IIFE pattern (immediately-invoked function expression):

```suggestion
    const float pushDir = [&]
    {
        if (curveDir == 0) {
            return ssvu::toRad(1.f) * -movement;
        }
       
        const int reverseDir = ((movement != 0) ? -movement : 1)*curveDir;
        return ssvu::toRad(1.f)*curveDir*reverseDir;
    }();
```

We are basically creating a lambda (anonymous function) and then immediately invoking it.

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 8]
```diff
@@ -19,35 +19,29 @@
...
-    : startPos{mStartPos}, pos{startPos}, hue{0}, angle{0},
-      size{Config::getPlayerSize()}, speed{Config::getPlayerSpeed()},
-      focusSpeed{Config::getPlayerFocusSpeed()}, dead{false}, swapTimer{36.f},
-      swapBlinkTimer{5.f}, deadEffectTimer{80.f, false}
-{
+CPlayer::CPlayer(const sf::Vector2f &mStartPos) noexcept
+        : startPos{mStartPos}, pos{startPos}, hue{0}, angle{0},
+          size{Config::getPlayerSize()}, speed{Config::getPlayerSpeed()},
+          focusSpeed{Config::getPlayerFocusSpeed()}, dead{false}, swapTimer{36.f},
+          swapBlinkTimer{5.f}, deadEffectTimer{80.f, false} {
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:30]: We need to initialize `lastPos` to `startPos` here.

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 9]
```diff
@@ -136,17 +127,70 @@ void CPlayer::drawDeathEffect(HexagonGame& mHexagonGame)
...
+        const int reverseDir = ((movement != 0) ? -movement : 1)*curveDir;
+        pushDir = ssvu::toRad(1.f)*curveDir*reverseDir;
+    } else {
+        pushDir = ssvu::toRad(1.f) * -movement;
+    }
+    while(wall.isOverlapping(pos)){
+        angle += pushDir;
+        pos = ssvs::getOrbitRad(startPos, angle, radius);
+
+        attempt++;
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:183]:
```suggestion
        ++attempt;
```

File: src/SSVOpenHexagon/Components/CPlayer.cpp (hunk 8 above)
Comment (SuperV1234) [reply src/SSVOpenHexagon/Components/CPlayer.cpp:30]: We also need to initialize `lastAngle`.

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 10]
```diff
@@ -188,9 +232,9 @@ void CPlayer::update(HexagonGame& mHexagonGame, ssvu::FT mFT)
 
     const sf::Vector2f tempPos{ssvs::getOrbitRad(startPos, angle, radius)};
     const sf::Vector2f pLeftCheck{
-        ssvs::getOrbitRad(tempPos, angle - ssvu::piHalf, 0.01f)};
+        ssvs::getOrbitRad(tempPos, angle - ssvu::piHalf, 1.01f)};
     const sf::Vector2f pRightCheck{
-        ssvs::getOrbitRad(tempPos, angle + ssvu::piHalf, 0.01f)};
+        ssvs::getOrbitRad(tempPos, angle + ssvu::piHalf, 1.01f)};
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:237]: Why was this changed?

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 11]
```diff
@@ -136,17 +127,70 @@ void CPlayer::drawDeathEffect(HexagonGame& mHexagonGame)
...
 
+[[nodiscard]] sf::Vector2f CPlayer::getPosition() const {
+    return pos;
+}
+
+[[nodiscard]] sf::Vector2f CPlayer::getStartPosition() const {
+    return startPos;
+}
+
+void CPlayer::kill(HexagonGame &mHexagonGame, bool mEffect) {
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:147]: What's the point of `mEffect`? Is there anywhere when this should be called with `mEffect=false`? Also, the name is not very self-explanatory. What "effect" are we talking about?

File: src/SSVOpenHexagon/Components/CWall.cpp [hunk 12]
```diff
@@ -62,15 +65,48 @@ void CWall::update(
...
+        if(moveCurve(mHexagonGame, mFT, vp, mCenterPos) && curve.speed != 0.f){
+            player.push(mHexagonGame, *this);
+        }
+    }
+
+    if(pointsOnCenter > 3)
+    {
+        killed = true;
+    }
+    */
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CWall.cpp:81]: Let's remove this commented code block

File: src/SSVOpenHexagon/Components/CWall.cpp [hunk 13]
```diff
@@ -62,15 +65,48 @@ void CWall::update(
...
+}
+
+bool CWall::moveTowardsCenter(HexagonGame& mHexagonGame, const sf::Vector2f& mCenterPos, ssvu::FT mFT){
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CWall.cpp:84]: I think this should just return `void`, and we should explicitly check for collision directly in `HexagonGame::updateWalls`. This function has too many responsibilities

File: src/SSVOpenHexagon/Components/CWall.cpp [hunk 14]
```diff
@@ -62,15 +65,48 @@ void CWall::update(
...
 
     if(pointsOnCenter > 3)
     {
         killed = true;
     }
+
+    return isOverlapping(mHexagonGame.player.getPosition());
+}
+
+bool CWall::moveCurve(HexagonGame& mHexagonGame, const sf::Vector2f& mCenterPos, ssvu::FT mFT){
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CWall.cpp:104]: Ditto, should return `void` and we should check collision directly in the loop in `HexagonGame::updateWalls`

File: src/SSVOpenHexagon/Core/HGUpdate.cpp [hunk 15]
```diff
@@ -266,6 +262,23 @@ void HexagonGame::update(ssvu::FT mFT)
...
+
+    for(CWall& wall : walls) {
+        player.push(*this, wall);
+
+        if (wall.moveTowardsCenter(*this, centerPos, mFT)) {
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Core/HGUpdate.cpp:272]: What I mean above is something like

```suggestion
        wall.moveTowardsCenter(*this, centerPos, mFT);
        if (wall.isOverlapping(player)) {
```

File: src/SSVOpenHexagon/Components/CPlayer.cpp (hunk 2 above)
Comment (Zly-u) [reply src/SSVOpenHexagon/Components/CPlayer.cpp:22]: It's my IDE that does this, didn't found a function to disable this auto formatting unfortunately.

File: src/SSVOpenHexagon/Online/Online.cpp (hunk 3 above)
Comment (Zly-u) [reply src/SSVOpenHexagon/Online/Online.cpp:519]: a hacky way of abusing multi-line comments. Very handy for testing and such, when you put another slash at the beginning of it while the end of the comment already has double slash - it will uncomment the thing and you don't have to get rid of the ending of it - `*/`
```cpp
/*
void main(){ return 0; }
//*/
```
```cpp
//*
void main(){ return 0; }
//*/
```

Comment (Zly-u) [reply src/SSVOpenHexagon/Online/Online.cpp:519]: Oh, but idk i didn't touch Online header though huh, well uh I thought that was my thing.

File: src/SSVOpenHexagon/Core/HGUpdate.cpp (hunk 15 above)
Comment (Zly-u) [reply src/SSVOpenHexagon/Core/HGUpdate.cpp:272]: alright, will change all of that soon.

File: src/SSVOpenHexagon/Components/CPlayer.cpp (hunk 10 above)
Comment (Zly-u) [reply src/SSVOpenHexagon/Components/CPlayer.cpp:237]: idk, was testing stuff and didn't notice any difference so I guess I left it for more precise collision detection...? Eh, idk honestly but this part will need to be changed actually dues to the custom walls being a complete whacks to properly calculate for collisions, so we need to leave it as just:

```cpp
if(wall.isOverlapping(pos)){
    kill(...);
}
```

File: src/SSVOpenHexagon/Components/CPlayer.cpp (hunk 11 above)
Comment (Zly-u) [reply src/SSVOpenHexagon/Components/CPlayer.cpp:147]: wanted to use that function to replace standard methods of killing s player to simplify it.
And by the effect it means the death effect itself... Prob had to name it like this.

Comment (Zly-u) [reply src/SSVOpenHexagon/Components/CPlayer.cpp:147]: also yes there are places where it supposed to be called with `mEffect=false` like for example while `tutorial mode` is on iirc.

Comment (Zly-u) [reply src/SSVOpenHexagon/Components/CPlayer.cpp:147]: also yes there are places where it supposed to be called with `mEffect=false` like for example while `tutorial mode` is on iirc.
